spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/outbound/PubSubMessageHandler.java

```diff
@@ -16,47 +16,28 @@
 
 package org.springframework.cloud.gcp.pubsub.integration.outbound;
 
-import com.google.cloud.pubsub.spi.v1.Publisher;
-import com.google.pubsub.v1.PubsubMessage;
-import com.google.pubsub.v1.TopicName;
 import java.io.IOException;
-import org.springframework.cloud.gcp.pubsub.integration.converters.SimpleMessageConverter;
+import org.springframework.cloud.gcp.pubsub.core.PubSubTemplate;
 import org.springframework.integration.handler.AbstractMessageHandler;
 import org.springframework.messaging.Message;
-import org.springframework.messaging.converter.MessageConversionException;
-import org.springframework.messaging.converter.MessageConverter;
 
 /**
- * Converts from internal Spring message to GCP Pub/Sub message and publishes the Pub/Sub
- * message to a Pub/Sub topic.
+ * Sends messages to Google Cloud Pub/Sub by delegating to {@link PubSubTemplate}.
  *
  * @author João André Martins
  */
 public class PubSubMessageHandler extends AbstractMessageHandler {
 
-  private Publisher publisher;
-  private MessageConverter messageConverter;
+  private PubSubTemplate pubsubTemplate;
+  private String topic;
 
-  public PubSubMessageHandler(String projectId, String topicName) throws IOException {
-    this(projectId, topicName, new SimpleMessageConverter());
-  }
-
-  public PubSubMessageHandler(String projectId, String topicName,
-      MessageConverter messageConverter) throws IOException {
-    publisher = Publisher.defaultBuilder(TopicName.create(
-        projectId, topicName)).build();
-    this.messageConverter = messageConverter;
+  public PubSubMessageHandler(String projectId, String topic) throws IOException {
+    pubsubTemplate = new PubSubTemplate(null, projectId);
+    this.topic = topic;
```

Comment:
We can't assume how complex is end-user flow and can't guess what is the source of metadata for incoming data. But any way your component (`PubSubMessageHandler`) should be ready for that and provide some flexibility and choice for end-user. That's why I always show that `KafkaProducerMessageHandler` as a sample with its flexible expressions. The same I can say about `KinesisMessageHandler`: https://github.com/spring-projects/spring-integration-aws/blob/master/src/main/java/org/springframework/integration/aws/outbound/KinesisMessageHandler.java#L93, where we have the logic like:
```
	String stream = message.getHeaders().get(AwsHeaders.STREAM, String.class);
		if (!StringUtils.hasText(stream) && this.streamExpression != null) {
			stream = this.streamExpression.getValue(this.evaluationContext, message, String.class);
		}
		Assert.state(stream != null, "'stream' must not be null for sending a Kinesis record. " +
				"Consider configuring this handler with a 'stream'( or 'streamExpression') or supply an " +
				"'aws_stream' message header.");

```